044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,275 @@
+
+the source will be configurable with a property like:
+
+1. `client.quota.callback.static.storage.volume.source=cluster`
+
+Local will be the default if the property isn't provided but is deprecated from the beginning.
+
+#### Cluster Volume Source
+
+With the introduction of [KIP-827](https://cwiki.apache.org/confluence/display/KAFKA/KIP-827%3A+Expose+log+dirs+total+and+usable+space+via+Kafka+API) in
+kafka 3.3 we can now obtain the total and usable (available) bytes per log dir as part of the DescribeLogDirsResponse.
+
+Note: if a single disk contains multiple log dirs, it will be described multiple times through the kafka APIs. This
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster.
+
+The performance cost of this API was [discussed](https://lists.apache.org/thread/11zyqqnyg1wgf4jdo6pvn7hn51g3vf8r) 
+upstream as part of the KIP, which should be low cost. The performance impact of polling is managed by making the poll
+interval configurable. The poll interval will be configurable using the existing `client.quota.callback.static.storage.check-interval` 
+property (default is 0 - disabled).
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions though the admin api we need to construct an admin client. We intend to
+connect to the replication listener and will piggyback on the broker's existing SSL configuration to
+obtain keystore/truststore properties.
+
+If using cluster sourced volumes we will expose these optional properties to customise the admin connection:
+
+1. `client.quota.callback.kafka.listener.name` (default `replication-9091`)
+2. `client.quota.callback.kafka.listener.port` (default `9091`)
+3. `client.quota.callback.kafka.listener.protocol` (default `SSL`)
+4. `client.quota.callback.kafka.clientIdPrefix` (default `__strimzi`)
+
+Setting `client.quota.callback.static.storage.hard` or `client.quota.callback.static.storage.soft` would be incompatible
+with the cluster sourced volumes and fail configuration of the plugin.
+
+The admin client will re-use the broker SSL configuration
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+used bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume.
+
+We retain the existing concept of a soft limit (where throttling begins) and a hard limit (where we effectively
```

Comment:
IIRC the 'soft' limit concept was borrowed from features common in the point to point messaging domain.  The idea there is to slow a producer so the consumers could 'catch up'.  consumers in the point to point domain are destructive - they cause messages to be removed from the broker as they are consumed - so free up disk space.   Kafka's model is completely different - disk space is only freed by compaction or retention.  Slowing a producer doesn't really make sense.

Reply:
I agree that soft limits are not as effective for Kafka topics as for point to point messaging, I think the idea of applying back pressure as the broker approaches disk limits still has utility. 

In scenarios like periodic jobs which produce large amounts of data in a burst. e.g. produce every 24hrs with 7 day expiry (one can argue about the merits of such a setup elsewhere) .  Throttling the batch job to 50% to allow the the topic cleaner to run is potentially easier to deal with than failed publication of specific messages and the resulting de-duplication on job replay etc.

Reply:
-0 from me.  To me the marginal benefit brought by this feature is outweighed by the extra cost of implementation, testing and documentation.

Reply:
I feel like there is a value in the _early warning_ from slowing down the clients. But I would tend to agree that it is not the core functionality and people can use alerting and other mechanisms as early warning as well. I do not know the code well enough to understand how much it will simplify it. But it sounds like something worth considering.

Reply:
If we were going to remove the soft limits, and the signal is implicitly provides, then I'd like for example alerting configuration to be provided and documented. Let's make it really easy for users to set this up _properly_.

Reply:
I thought I had said this before but, soft limits are very simple to implement. There would be minimal savings in removing them from the implementation, there is a bit more from removing the properties and documentation. 

I'm still of the view they add more value in general than they cost.

Reply:
We're going with the majority sentiment of a mildly negative feeling about the soft limits and have removed them (noting we could still add them in future if desired).